### PR TITLE
RavenDB-8832: Implement and use PeepingTomStream

### PR DESCRIPTION
**Pull request by aviviadi**

### CODE REVIEW

**Comment by aviviadi** [comment]
@ayende - please wait until I add ReadAsync too

File: src/Raven.Client/Documents/Commands/MultiGet/MultiGetCommand.cs [hunk 1]
```diff
@@ -119,19 +120,20 @@ public override void SetResponseRaw(HttpResponseMessage response, Stream stream,
             using (var parser = new UnmanagedJsonParser(context, state, "multi_get/response"))
             using (context.GetManagedBuffer(out JsonOperationContext.ManagedPinnedBuffer buffer))
             {
-                if (UnmanagedJsonParserHelper.Read(stream, parser, state, buffer) == false)
-                    ThrowInvalidResponse();
+                var peepingTomStream = new PeepingTomStream(stream);
```
Comment (ayende) [inline src/Raven.Client/Documents/Commands/MultiGet/MultiGetCommand.cs:123]: This generates a 4KB buffer on every MultIGet call, let's see if we can use the context to get a pinnable buffer that is reused between calls.

File: src/Sparrow/Json/Parsing/UnmanagedJsonParserHelper.cs [hunk 2]
```diff
@@ -54,19 +57,36 @@ public static bool Read(Stream stream, UnmanagedJsonParser parser, JsonParserSta
...
+        public static void ThrowInvalidJson(PeepingTomStream peepingTomStream)
         {
-            throw new InvalidOperationException("Invalid JSON.");
+            string s = GetPeepingTomBufferAsString(peepingTomStream);
+
+            throw new InvalidOperationException("Invalid JSON: " + s);
         }
 
-        public static void ReadObject(BlittableJsonDocumentBuilder builder, Stream stream, UnmanagedJsonParser parser, JsonOperationContext.ManagedPinnedBuffer buffer)
+        private static string GetPeepingTomBufferAsString(PeepingTomStream peepingTomStream)
```
Comment (ayende) [inline src/Sparrow/Json/Parsing/UnmanagedJsonParserHelper.cs:67]: Duplicated code from RavenCommand.ThrowInvalidJson

File: src/Sparrow/PeepingTomStream.cs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+            {
+                start = 0;
+                size = _pos;
+            }
+            else
+            {
+                start = _pos;
+                size = BufferWindowSize;
+            }
+            while ((_bufferWindow[start] & 0x80) != 0)
```
Comment (ayende) [inline src/Sparrow/PeepingTomStream.cs:69]: Need to add a comment here to explain it.

**Review by ayende — CHANGES_REQUESTED** [review]

**Comment by ayende** [comment]
Okay, closing until then

**Pull request closed without merging**